syntax = "proto3";

import "google/protobuf/empty.proto";
import "minecraft/requests.proto";
import "minecraft/updates.proto";

package minecraft;

service World {
    rpc ModifyWorld (WorldActionRequest) returns (Errors);
    rpc WorldUpdates (ClientData) returns (stream WorldUpdate);

    rpc MetadataUpdates (google.protobuf.Empty) returns (stream Metadata);
}

message Errors {
    string error_message = 1;
}

message ClientData {
    string name = 1;
    bool send_existing_state = 2;
}

message Metadata {
    // unique identifier -> client name
    map<string, string> clients = 1;
    uint64 total_blocks = 2;
}
